podresources: do not export empty NUMA topology

AlexeyPerevalov opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
If device plugin returns device without topology, keep it internaly
as NUMA node -1, it helps at podresources level to not export NUMA
topology, otherwise topology is exported with NUMA node id 0,
which is not accurate.


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:
It's impossible to unveile this bug just by tracing json.Marshal(resp)
in podresource client, because NUMANodes field ID has json property
omitempty, in this case when ID=0 shown as emtpy NUMANode.
To reproduce it, better to iterate on devices and just
trace dev.Topology.Nodes[0].ID.
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Commit:
podresource: do not export NUMA topology if it's empty

If device plugin returns device without topology, keep it internaly
as NUMA node -1, it helps at podresources level to not export NUMA
topology, otherwise topology is exported with NUMA node id 0,
which is not accurate.

It's imposible to unveile this bug just by tracing json.Marshal(resp)
in podresource client, because NUMANodes field ID has json property
omitempty, in this case when ID=0 shown as emtpy NUMANode.
To reproduce it, better to iterate on devices and just
trace dev.Topology.Nodes[0].ID.

Signed-off-by: Alexey Perevalov <alexey.perevalov@huawei.com>

fromanirh commented:
/priority important-longterm
/triage accepted

Do we need a corresponding issue?
This probably calls for a e2e test (or just an extension of the existing ones).

fromanirh commented:
/cc

AlexeyPerevalov commented:
> /priority important-longterm
> /triage accepted
> 
> Do we need a corresponding issue?
> This probably calls for a e2e test (or just an extension of the existing ones).

I think yes, we can create an issue. I'll do it.
It's a complex e2e test, it includes:
  1. device plugin which report devices w/o topology
  2. test pod, consuming these resources
  3. kubelet restart
  4. check for result
 I didn't find such, but it worth adding it.

swatisehgal commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -344,18 +344,20 @@ func (pdev *podDevices) getContainerDevices(podUID, contName string) ResourceDev
-					if nodes := pDev.Topology.GetNodes(); nodes != nil {
-						NUMANodes = append(NUMANodes, nodes...)
+				var topology *pluginapi.TopologyInfo
+				if numaid != nodeWithoutTopology {
```
I think rather than skipping devices that don't have topology we should expose NUMA ID as -1 to the podresource API client that there are devices without topology and let the client decide what it would like to do with this information. This might need some additional documentation but I think just simply skipping the devices is not the best approach because it gives the impression that devices with the Topology are the only allocated devices which might lead to incorrect accounting of resources if that is the goal of the client.

AlexeyPerevalov replied:
device is not skipped, it will be there, but with empty Topology field.

swatisehgal replied:
Ah right! Sorry I misread. Looks good!

swatisehgal reviewed (commented):
Thanks for the PR! Please take a look at my suggestion in the comment below.

swatisehgal commented:
+1 for the e2e test!

ehashman commented:
/release-note-none

waiting on e2e test

AlexeyPerevalov commented:
> 
> 
> > It's a complex e2e test, it includes:
> > 
> > 1. device plugin which report devices w/o topology
> > 2. test pod, consuming these resources
> > 3. restart of kubelet
> > 4. check for result
> >    I didn't find such, but it worth adding it.
> 
> +1 for the e2e test!

Currently test/e2e_node/podresources_test.go uses docker.io/nfvpe/sriov-device-plugin:, it's probably
https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin, which creates topology always, even there is no NUMA topology on the host (https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/blob/1f1822bf0bbb25bff55190fcad861617c1a2abb7/pkg/resources/pciDevice.go#L81)
I don't think this device plugin is suitable for e2e test for case of this fix.
E.g. github.com/k8stopologyawareschedwg/sample-device-plugin/ which provides predefined Device set w/o topology for any hw environment where it runs, fits much better. But I heart mocks are prohibited for e2e testing, only existing plugins allowed.

 https://github.com/kubevirt/kubernetes-device-plugin - also fits, since it doesn't fil topology field and it uses podresources v1beta1.
The question is which device plugin is better?

swatisehgal commented:
@AlexeyPerevalov This item is relevant not only for this feature but also for other E2E tests because when tests run on nodes without SR-IOV NICs they are skipped and the . I believe @fromanirh intends to have a discussion about how the CI experience can be improved and that should shape the way we proceed with E2E test for this PR.

AlexeyPerevalov commented:
/hold still checking e2e_node tests

fromanirh commented on test/e2e_node/podresources_test.go:
```diff
@@ -644,9 +675,85 @@ var _ = SIGDescribe("POD Resources [Serial] [Feature:PodResources][NodeFeature:P
 			ginkgo.By("checking GetAllocatableResources fail if the feature gate is not enabled")
-			_, err = cli.GetAllocatableResources(context.TODO(), &kubeletpodresourcesv1.AllocatableResourcesRequest{})
+			allocatableRes, err := cli.GetAllocatableResources(context.TODO(), &kubeletpodresourcesv1.AllocatableResourcesRequest{})
+			framework.Logf("GetAllocatableResources result: %v, err: %v", allocatableRes, err)
```
this change is for debug only, right?

AlexeyPerevalov replied:
yes, do you want to keep it as it was before?

fromanirh replied:
no, I don't mind too much. It is actually a bit better like this, so let's keep it like you wrote.

fromanirh commented on test/e2e_node/podresources_test.go:
```diff
@@ -644,9 +675,85 @@ var _ = SIGDescribe("POD Resources [Serial] [Feature:PodResources][NodeFeature:P
+
+			ginkgo.By("checking List and resources kubevirt resource should be without topology")
+
+			allocatableResponse, _ := cli.GetAllocatableResources(context.TODO(), &kubeletpodresourcesv1.AllocatableResourcesRequest{})
```
I'd check `framework.ExpectNoError(err)` here.

ffromani reviewed (commented):
thanks for adding a new e2e test so comprehensive. Adding the kubevirt plugin will surely help.
I think this PR is good, only two things are missing:
1. if I'm not mistaken, the kubevirt device plugin was deprecated and it is now kept for backward compatibility. We need to make sure the image stays around long enough, or we need a replacement (or a safe mirror we can use in the future)
2. I'd like to run the e2e tests myself and play with them for a little while
ofc point 1 above is more important.

fromanirh commented:
>     1. if I'm not mistaken, the kubevirt device plugin was deprecated and it is now kept for backward compatibility. We need to make sure the image stays around long enough, or we need a replacement (or a safe mirror we can use in the future)

I asked in the virtualization channel: https://kubernetes.slack.com/archives/C8ED7RKFE/p1632742391177200
Worst case we can mirror that image, it should not be a big deal.

>     2. I'd like to run the e2e tests myself and play with them for a little while
>        ofc point 1 above is more important.

On it now

fromanirh commented:
/test pull-kubernetes-typecheck

fromanirh commented on test/e2e_node/podresources_test.go:
```diff
@@ -169,7 +174,27 @@ func findContainerDeviceByName(devs []*kubeletpodresourcesv1.ContainerDevices, r
-func matchPodDescWithResources(expected []podDesc, found podResMap) error {
+type deviceCheckFunc func(expect podDesc, found podResMap) error
+
+func checkForTopology(expect podDesc, found podResMap) error {
```
I'd just inline this check in expectPodResources.
As it is now, if the kubevirt resource name is not found, we silently skip and succeed. Then it makes sense to just always run this check, avoiding the extra argument.

Once the inlining is done, we should add another check: if the resource is not found *but it was expected*, we should fail and error out.

To learn if the resource was expected, we need to make logic, and possibly add a field, to `podDesc`.

fromanirh replied:
I'd inline the check around here:
```
			dev := findContainerDeviceByName(cntInfo.GetDevices(), podReq.resourceName)
			if dev == nil {
			if dev == nil {
				return fmt.Errorf("pod %q container %q expected data for resource %q not found", podReq.podName, podReq.cntName, podReq.resourceName)
				return fmt.Errorf("pod %q container %q expected data for resource %q not found", podReq.podName, podReq.cntName, podReq.resourceName)
			}
			}
			if len(dev.DeviceIds) != podReq.resourceAmount {
			if len(dev.DeviceIds) != podReq.resourceAmount {
				return fmt.Errorf("pod %q container %q resource %q expected %d items got %v", podReq.podName, podReq.cntName, podReq.resourceName, podReq.resourceAmount, dev.DeviceIds)
				return fmt.Errorf("pod %q container %q resource %q expected %d items got %v", podReq.podName, podReq.cntName, podReq.resourceName, podReq.resourceAmount, dev.DeviceIds)
			}
			}
```
at the bottom of this `if`.

ffromani reviewed (commented):
I only have a comment about the e2e test implementation, everything else LGTM, and kudos for adding the KVM device plugin from kubevirt. This will turn very handy for future tests.

fromanirh commented:
/lgtm
after some extra thought, this PR is good enough, we can do some extra polishing later, I'm planning a cleanup/test extension PR anyway.

ehashman commented:
/hold cancel
appears to have been left on since Aug.

swatisehgal commented:
/assign @klueska 

for node approval

klueska commented:
The code changes in the `kubelet` code look OK.

I didn't review the tests thoroughly, but I did see some things I normally don't like to approve (including commented out code blocks with no explanation). @fromanirh do you plan to include these tests as part of your cleanup effort?

fromanirh commented:
/lgtm cancel
I realized I automatically skipped https://github.com/kubernetes/kubernetes/pull/103289/commits/2a9da5d6b1ec443a3e10ab809e39512938083bd8 which is obviously a debug aid and I assumed was removed already. Apologies for that.

fromanirh commented:
> The code changes in the `devicemanger` look OK.
> 
> I didn't review the tests thoroughly, but I did see some things I normally don't like to approve (including commented out code blocks with no explanation). @fromanirh do you plan to include these tests as part of your cleanup effort?

Thanks Kevin for the wakeup call, I withdrawn my LGTM.

- The test changes in https://github.com/kubernetes/kubernetes/pull/103289/commits/bb81101570c9ef584e37a1fa7d079e314bcb4ce8	 (aka the first commit in this PR) looks good enough and even though `podresources_test.go` can use some improvement which I'd like to do later, I think they are sufficient. In general the whole test suite can use some simplification, so better to do everything in one go and in one separate PR.
- The test changes in https://github.com/kubernetes/kubernetes/pull/103289/commits/2a9da5d6b1ec443a3e10ab809e39512938083bd8 ( aka the second commit in this PR) are obviously a debug helper to understand the serial lane failure and can't be merged indeed.

swatisehgal committed:
podresource: move the checkForTopology logic inline

As per the recommendation here: https://github.com/kubernetes/kubernetes/pull/103289#pullrequestreview-766949859
we move the check inline.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

fromanirh commented:
/retest
